substrate/frame/nomination-pools/src/lib.rs

```diff
@@ -2995,6 +2995,84 @@ pub mod pallet {
 			Ok(Pays::No.into())
 		}
 
+		/// Reap the member from the pool. Effective only if their contribution to the pool falls
+		/// below ED.
+		#[pallet::call_index(26)]
+		// FIXME eagr weight functions
+		// #[pallet::weight(T::WeightInfo::reap_member())]
+		#[pallet::weight(1)]
+		pub fn reap_member(
+			origin: OriginFor<T>,
+			member_account: AccountIdLookupOf<T>,
+			num_slashing_spans: u32,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			let member = T::Lookup::lookup(member_account)?;
+			ensure!(
+				!Self::api_member_needs_delegate_migration(member.clone()),
+				Error::<T>::NotMigrated,
+			);
+
+			let (mut pool_member, mut bonded_pool, mut reward_pool) =
+				Self::get_member_with_pools(&member)?;
+			let contribution = pool_member.active_balance();
+
+			if contribution < T::Currency::minimum_balance() {
+				let pool_id = pool_member.pool_id;
+
+				reward_pool.update_records(
+					pool_id,
+					bonded_pool.points,
+					bonded_pool.commission.current(),
+				)?;
+				let _ = Self::do_reward_payout(
+					&member,
+					&mut pool_member,
+					&mut bonded_pool,
+					&mut reward_pool,
+				);
+
+				if contribution > Zero::zero() {
+					T::StakeAdapter::member_dust(
+						Member::from(member.clone()),
+						Pool::from(bonded_pool.bonded_account()),
+						contribution,
+						num_slashing_spans,
+					)?;
+
+					// FIXME eagr is this needed?
+					// Self::deposit_event(Event::<T>::Withdrawn {
+					// 	member: member.clone(),
+					// 	pool_id,
+					// 	points: bonded_pool.balance_to_point(contribution),
+					// 	balance: contribution,
+					// });
+				}
+
+				ClaimPermissions::<T>::remove(&member);
+				PoolMembers::<T>::remove(&member);
+
+				Self::deposit_event(Event::<T>::MemberRemoved {
+					pool_id,
+					member: member.clone(),
+					// FIXME eagr there should not be any dangling delegation at this point right?
+					released_balance: Zero::zero(),
+				});
+
+				if member == bonded_pool.roles.depositor {
+					Pallet::<T>::dissolve_pool(bonded_pool);
+				} else {
+					let sub_pools = SubPoolsStorage::<T>::get(pool_id).unwrap_or_default();
+					bonded_pool.dec_members().put();
+					SubPoolsStorage::<T>::insert(pool_id, sub_pools);
+				}
+			}
+
+			Ok(())
+
+			// FIXME eagr should return fee info?
```

Comment:
Really depends on what you want to do.
For permissionless extrinsic (that anyone can call to cleanup chain if some criteria is fulfilled), we want to not charge any fees. So if the call succeeds, we should refund the fees.

See https://docs.substrate.io/build/tx-weights-fees/#post-dispatch-weight-correction.

Though, I believe we might not need this to be an [extrinsic anyway](https://github.com/paritytech/polkadot-sdk/issues/5009#issuecomment-2382738660)?